# Clustering ICD9 Codes

In [1]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
import numpy as np
np.set_printoptions(threshold=np.nan)
from sklearn import random_projection
from kmodes.kmodes import KModes

In [2]:
df = pd.read_csv('DIAGNOSES_ICD.csv',usecols=['HADM_ID','ICD9_CODE'])
df['ICD9_CODE'] = df['ICD9_CODE'].astype(str)
#df = df.head(1000)

In [3]:
le = LabelEncoder()
le.fit(df['ICD9_CODE'])
le_name_mapping = dict(zip(le.classes_, le.transform(le.classes_)))
#print(le_name_mapping)

labelencoder = LabelEncoder()
df['ICD9_CODE'] = labelencoder.fit_transform(df['ICD9_CODE'])
#print(df)

{'0030': 0, '0031': 1, '0038': 2, '0039': 3, '0041': 4, '0048': 5, '0049': 6, '0051': 7, '00581': 8, '0059': 9, '0071': 10, '0074': 11, '00804': 12, '00841': 13, '00843': 14, '00845': 15, '00847': 16, '0085': 17, '00861': 18, '00862': 19, '00863': 20, '00869': 21, '0088': 22, '0090': 23, '0091': 24, '0092': 25, '0093': 26, '01085': 27, '01123': 28, '01136': 29, '01164': 30, '01186': 31, '01190': 32, '01193': 33, '01194': 34, '01205': 35, '01215': 36, '01300': 37, '01304': 38, '01325': 39, '01330': 40, '01354': 41, '01402': 42, '01405': 43, '01485': 44, '01504': 45, '01505': 46, '01803': 47, '01805': 48, '01880': 49, '01890': 50, '01894': 51, '01895': 52, '01896': 53, '0218': 54, '0239': 55, '0270': 56, '0272': 57, '0279': 58, '0309': 59, '0310': 60, '0311': 61, '0312': 62, '0318': 63, '0319': 64, '03285': 65, '03289': 66, '0329': 67, '0338': 68, '0340': 69, '035': 70, '0360': 71, '0362': 72, '03641': 73, '0380': 74, '03810': 75, '03811': 76, '03812': 77, '03819': 78, '0382': 79, '0383'

In [4]:
patient_list = []
for index, row in df.iterrows():
    if(row['ICD9_CODE'] == 824):
        patient_list.append(row['HADM_ID'])
#print(patient_list)

In [5]:
Patient_lib = {}
for index, row in df.iterrows():
    code = row['ICD9_CODE']
    patient = row['HADM_ID']
    if(patient in patient_list):
        Patient_lib.setdefault(patient, []).append(code)
#print(Patient_lib)

{140784: [2861, 4540, 184, 1104, 2109, 1077, 988, 1962, 824], 164853: [2853, 179, 2856, 2957, 2869, 2498, 1092, 2537, 1104, 88, 2451, 1035, 921, 984, 995, 5873, 1277, 1962, 2868, 824, 807, 247], 195632: [1992, 2019, 1962, 824, 1309], 113323: [5968, 5963, 2530, 5935, 2098, 3817, 824, 2633, 2018, 6777], 199634: [3677, 290, 2482, 686, 6871, 3709, 2633, 1962, 824], 112086: [2063, 5937, 2526, 3028, 5934, 1632, 1962, 943, 824, 2633, 2755, 6651], 158569: [2144, 1958, 1952, 1974, 3501, 2633, 943, 824, 2030], 143045: [2019, 2012, 824, 1962, 943], 157681: [2019, 2012, 824, 946, 1962], 109451: [2004, 4450, 2922, 2957, 1974, 2109, 2352, 2487, 2098, 2096, 2019, 824, 1078, 247, 943, 967, 6561, 2199], 111970: [87, 4451, 1974, 2098, 3495, 2501, 3395, 5968, 15, 2865, 5873, 6537, 824, 1083, 2199, 967, 2019, 476, 2245, 807, 6204], 161859: [2000, 2019, 2633, 824], 162569: [2019, 2012, 2482, 2015, 824, 943, 2083, 1249, 6651, 6778, 6676], 166707: [1951, 2077, 1104, 5935, 2098, 2099, 2019, 1962, 807, 824, 35

In [65]:
hot_df = pd.read_csv('hot_df.csv')
'''
hot_df = pd.DataFrame()
for patient, code_list in Patient_lib.items():
    patient_codes = {}
    for code in code_list:
        patient_codes[code] = 1
    hot_df = hot_df.append(patient_codes,ignore_index=True)
    print(patient)
hot_df.to_csv('hot_df.csv')
'''
print(hot_df.shape)

(9057, 3515)


In [66]:
hot_df.fillna(0)
hot_df = hot_df.drop(columns=['patients'])
code_list = list(hot_df)
for code in code_list:  
    hot_df.at['Total',code] = hot_df[code].sum()
    #print(code)
#print(hot_df)

184
824
988
1077
1104
1962
2109
2861
4540
88
179
247
807
921
984
995
1035
1092
1277
2451
2498
2537
2853
2856
2868
2869
2957
5873
1309
1992
2019
2018
2098
2530
2633
3817
5935
5963
5968
6777
290
686
2482
3677
3709
6871
943
1632
2063
2526
2755
3028
5934
5937
6651
1952
1958
1974
2030
2144
3501
2012
946
967
1078
2004
2096
2199
2352
2487
2922
4450
6561
15
87
476
1083
2245
2501
2865
3395
3495
4451
6204
6537
2000
1249
2015
2083
6676
6778
1951
2077
2099
3584
2107
2694
6550
6758
2020
2120
2349
2373
2471
4569
5958
1524
1973
2156
2469
2848
2965
4436
359
536
2017
2112
2347
2780
2961
3038
5957
5959
6350
6605
6808
503
896
1662
3398
6337
991
1512
2334
2338
2536
2860
105
897
985
1420
1466
2715
2923
2964
3700
3801
3882
4542
4555
5922
5977
6197
6380
6548
6805
99
2327
2330
4829
5938
6120
6228
6873
2778
5874
6730
1373
3713
83
2981
5872
2062
2072
2474
983
1160
2207
2953
3720
5914
5990
6371
2111
2148
2200
2266
5912
1093
1586
2114
2183
2203
2115
4368
6206
3852
6757
996
2176
2178
2434
3822
4390
4482
6533
6864


736
6574
4809
4816
4818
5122
2331
4383
6329
52
48
6899
1766
1430
1516
1831
4142
4388
4485
3635
4897
3410
3688
949
1520
5179
6427
2008
3574
3681
1986
96
2260
1188
1908
1615
1535
3856
4672
3613
2085
4524
2528
6639
203
1544
2967
2045
3729
6333
2382
451
1894
4678
2435
6972
2370
3508
1339
2312
5353
6310
5084
738
2129
2683
4576
2574
4434
1326
1390
4424
3823
384
5149
3012
2339
6745
2160
2324
5664
5667
5486
993
3429
2513
4619
3873
4701
4800
6068
1934
2264
3551
1983
1075
6543
5507
5739
1558
1362
5142
4790
6562
1022
553
6849
364
2491
1140
2705
488
2619
2440
350
2196
5907
1131
6799
6912
1137
1769
6439
6901
2885
5092
1861
6398
4992
2785
1700
3889
3891
6862
1784
1806
5185
524
3655
6834
5940
2031
2292
5588
4827
4885
5100
5290
5317
6833
1748
1636
3393
741
2304
6965
3793
3535
5307
6896
1296
2573
2127
2768
5962
6861
6365
4495
5529
5533
6471
6976
6560
6964
4492
878
2630
1342
6138
2003
3094
3136
2293
3850
5443
6048
1856
4562
1540
3680
4724
2412
2313
6728
691
2754
523
1459
2091
3440
1031
1811
1921
5928
65

In [67]:
#print(hot_df)
for code in code_list:
    if((hot_df.at['Total',code]) < 100):
        hot_df = hot_df.drop(columns=[code])
        #print(code)
print(hot_df.shape)

         184     824    988   1077   1104    1962    2109   2861  4540     88  \
0        1.0     1.0    1.0    1.0    1.0     1.0     1.0    1.0   1.0    NaN   
1        NaN     1.0    NaN    NaN    1.0     1.0     NaN    NaN   NaN    1.0   
2        NaN     1.0    NaN    NaN    NaN     1.0     NaN    NaN   NaN    NaN   
3        NaN     1.0    NaN    NaN    NaN     NaN     NaN    NaN   NaN    NaN   
4        NaN     1.0    NaN    NaN    NaN     1.0     NaN    NaN   NaN    NaN   
5        NaN     1.0    NaN    NaN    NaN     1.0     NaN    NaN   NaN    NaN   
6        NaN     1.0    NaN    NaN    NaN     NaN     NaN    NaN   NaN    NaN   
7        NaN     1.0    NaN    NaN    NaN     1.0     NaN    NaN   NaN    NaN   
8        NaN     1.0    NaN    NaN    NaN     1.0     NaN    NaN   NaN    NaN   
9        NaN     1.0    NaN    NaN    NaN     NaN     1.0    NaN   NaN    NaN   
10       NaN     1.0    NaN    NaN    NaN     NaN     NaN    NaN   NaN    NaN   
11       NaN     1.0    NaN 

4540
179
247
1092
1277
2498
2869
1992
3817
5963
290
686
3677
3709
1632
2755
5934
5937
1952
1958
3501
2352
87
476
2865
3395
3495
6537
1249
2083
6676
1951
2694
2120
2373
2471
4569
1524
2848
4436
359
536
3038
5959
503
896
1662
6337
1512
2334
2338
2860
105
897
1420
1466
2715
3700
3801
3882
5922
5977
6197
6380
6548
6805
2327
2330
4829
6228
2778
5874
6730
3713
2981
1160
2207
3720
5990
6371
2111
2148
2200
2266
1586
4368
3852
2176
2178
3822
4390
4482
6864
2005
2204
774
1069
3714
6215
6565
86
339
2049
2484
5879
2009
3488
3493
3811
957
1273
2613
2628
2882
3676
3790
6578
978
4426
1527
1719
1765
2106
6607
664
5626
6270
3941
2209
2251
3699
6760
6802
2110
370
1560
2863
3507
250
619
1062
2795
977
2325
2328
2348
2794
75
2381
2433
3774
2336
2877
4607
1043
4592
2025
2061
2028
2082
6957
2525
2453
2201
2713
4510
5944
6820
6909
579
2310
2545
2548
2635
891
2095
2888
5875
296
3500
2122
2426
2452
5036
6312
1575
2620
2693
2920
5758
5782
5783
6124
6132
6133
2040
3523
3777
1102
1574
3394
6115
120
1465
107
609
34

4475
1833
4913
5569
92
5508
27
1055
2185
736
6574
4809
4816
4818
5122
2331
4383
6329
52
48
6899
1766
1430
1516
1831
4142
4388
4485
3635
4897
3410
3688
949
1520
5179
6427
2008
3574
3681
1986
96
2260
1188
1908
1615
1535
3856
4672
3613
2085
4524
2528
6639
203
1544
2967
2045
3729
6333
2382
451
1894
4678
2435
6972
2370
3508
1339
2312
5353
6310
5084
738
2129
2683
4576
2574
4434
1326
1390
4424
3823
384
5149
3012
2339
6745
2160
2324
5664
5667
5486
993
3429
2513
4619
3873
4701
4800
6068
1934
2264
3551
1983
1075
6543
5507
5739
1558
1362
5142
4790
6562
1022
553
6849
364
2491
1140
2705
488
2619
2440
350
2196
5907
1131
6799
6912
1137
1769
6439
6901
2885
5092
1861
6398
4992
2785
1700
3889
3891
6862
1784
1806
5185
524
3655
6834
5940
2031
2292
5588
4827
4885
5100
5290
5317
6833
1748
1636
3393
741
2304
6965
3793
3535
5307
6896
1296
2573
2127
2768
5962
6861
6365
4495
5529
5533
6471
6976
6560
6964
4492
878
2630
1342
6138
2003
3094
3136
2293
3850
5443
6048
1856
4562
1540
3680
4724
2412
2313
6728
691
2754


In [68]:
hot_df.to_csv('min_hot_df.csv')

In [77]:
min_hot_df = pd.read_csv('min_hot_df.csv')

In [78]:
min_hot_df.fillna(0)
min_hot_df = min_hot_df.drop(columns=['patients'])
code_list = list(min_hot_df)
for code in code_list:
    if((min_hot_df.at[9057,code]) < 200):
        min_hot_df = min_hot_df.drop(columns=[code])
        #print(code)
print(min_hot_df.shape)

184
988
921
1035
2537
2853
2856
6871
2199
2922
2000
2077
3584
6550
2020
2156
2469
2112
2780
2961
6350
3398
4542
99
5938
83
5914
5912
1093
2183
2203
6206
996
2434
6533
4379
2996
1175
1438
1080
5902
2797
1582
6747
4493
2867
2100
6642
4496
2916
103
3036
6746
992
3562
2158
6743
108
1637
6675
2512
2910
1585
2539
986
2036
1503
5960
2353
5941
1569
2503
6244
1955
6237
6922
1588
6546
2792
2540
1120
6738
6787
1737
2222
4366
6744
3505
1601
6783
6929
6761
1251
1589
2547
6198
4483
(9058, 105)


In [80]:
min_hot_df.to_csv('min_hot_df.csv')

In [85]:
min_hot_df = pd.read_csv('min_hot_df.csv')
#print(min_hot_df)
min_hot_df = min_hot_df.drop(columns=['patients','824'])

       1077   1104    1962    2109   2861     88     807    984    995   2451  \
0       1.0    1.0     1.0     1.0    1.0    NaN     NaN    NaN    NaN    NaN   
1       NaN    1.0     1.0     NaN    NaN    1.0     1.0    1.0    1.0    1.0   
2       NaN    NaN     1.0     NaN    NaN    NaN     NaN    NaN    NaN    NaN   
3       NaN    NaN     NaN     NaN    NaN    NaN     NaN    NaN    NaN    NaN   
4       NaN    NaN     1.0     NaN    NaN    NaN     NaN    NaN    NaN    NaN   
5       NaN    NaN     1.0     NaN    NaN    NaN     NaN    NaN    NaN    NaN   
6       NaN    NaN     NaN     NaN    NaN    NaN     NaN    NaN    NaN    NaN   
7       NaN    NaN     1.0     NaN    NaN    NaN     NaN    NaN    NaN    NaN   
8       NaN    NaN     1.0     NaN    NaN    NaN     NaN    NaN    NaN    NaN   
9       NaN    NaN     NaN     1.0    NaN    NaN     NaN    NaN    NaN    NaN   
10      NaN    NaN     NaN     NaN    NaN    NaN     1.0    NaN    NaN    NaN   
11      NaN    NaN     NaN  

In [86]:
#print(hot_df.shape)
np_min_hot_df = min_hot_df.values
np_min_hot_df = np.nan_to_num(np_min_hot_df)
#transformer = random_projection.SparseRandomProjection(eps=.4)
#new_hot_df = transformer.fit_transform(np_hot_df)
print(np_min_hot_df.shape)


(9058, 104)


In [89]:
km = KModes(n_clusters=4, init='Cao',n_init=5,verbose=1)
clusters=km.fit_predict(np_min_hot_df)

Initialization method and algorithm are deterministic. Setting n_init to 1.
Init: initializing centroids
Init: initializing clusters
Starting iterations...
Run 1, iteration: 1/100, moves: 1416, cost: 59968.0
Run 1, iteration: 2/100, moves: 216, cost: 59926.0
Run 1, iteration: 3/100, moves: 11, cost: 59926.0


In [90]:
print(km.cluster_centroids_)

[[0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 0. 1. 0. 0. 0. 0. 0. 1. 1. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 0. 1. 0. 0. 0. 1. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0.
  0. 0. 0. 0. 0. 1. 1. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 1. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 0. 0. 0. 0. 0. 0.

icd9code:
1st cluster - 1962 -> 401.9 - unspecified essential hypertension

2nd cluster - 2109 -> 428.1 - left heart failure
              88   -> 038.9 - unspecified septicemia
              2957 -> 584.9 - acute kidney failure, unspecified
              5873 -> 995.92 - severe sepsis -> caused by sig deop in bp
              2098 -> 427.31 - atrail fibrillation -> irregular heartbeat
              4451 -> 785.52 - septic shock -> symptom low bp